ptz-device.cpp

```diff
@@ -343,16 +362,49 @@ void ptz_load_devices()
 					Q_ARG(uint32_t, calldata_int(cd, "device_id")),
 					Q_ARG(int, calldata_int(cd, "preset_id")));
 	};
-	proc_handler_add(ph, "void ptz_preset_recall(int device_id, int preset_id)",
+	proc_handler_add(ptz_ph, "void ptz_preset_recall(int device_id, int preset_id)",
 			ptz_preset_recall, NULL);
 
-	/* Pantilt Callback */
-	auto ptz_pantilt = [](void *data, calldata_t *cd) {
-		QMetaObject::invokeMethod(&ptzDeviceList, "pantilt",
-					Q_ARG(uint32_t, calldata_int(cd, "device_id")),
-					Q_ARG(double, calldata_float(cd, "pan")),
-					Q_ARG(double, calldata_float(cd, "tilt")));
+	auto ptz_move_continuous = [](void *data, calldata_t *cd) {
+		long long device_id;
+		double pan, tilt, zoom, focus;
+		if (!calldata_get_int(cd, "device_id", &device_id))
+			return;
+		int flags = 0;
+		if (calldata_get_float(cd, "pan", &pan) && calldata_get_float(cd, "tilt", &tilt))
+			flags |= MOVE_FLAG_PANTILT;
+		if (calldata_get_float(cd, "zoom", &zoom))
+			flags |= MOVE_FLAG_ZOOM;
+		if (calldata_get_float(cd, "focus", &focus))
+			flags |= MOVE_FLAG_FOCUS;
+		QMetaObject::invokeMethod(&ptzDeviceList, "move_continuous",
+						Q_ARG(uint32_t, device_id),
+						Q_ARG(uint32_t, flags),
+						Q_ARG(double, pan), Q_ARG(double, tilt),
+						Q_ARG(double, zoom),
+						Q_ARG(double, focus));
 	};
-	proc_handler_add(ph, "void ptz_pantilt(int device_id, float pan, float tilt)",
-			ptz_pantilt, NULL);
+	proc_handler_add(ptz_ph, "void ptz_move_continuous(int device_id, float pan, float tilt, float zoom, float focus)",
+			ptz_move_continuous, NULL);
+
+	/* Register the new proc hander with the main proc handler */
+	proc_handler_t *ph = obs_get_proc_handler();
+	if (!ph)
+		return;
+
+	/* Register a function for retrieving the PTZ call handler */
+	auto ptz_get_proc_handler = [](void *data, calldata_t *cd) {
+		calldata_set_ptr(cd, "return", ptz_ph);
+	};
+	proc_handler_add(ph, "ptr ptz_get_proc_handler()", ptz_get_proc_handler, NULL);
```

Comment:
The correct way to phrase the proc name here would actually be `void ptz_get_proc_handler(out ptr return)`

Reply:
Ah, okay. I had read the proc_handler code and saw that would parse a return type in this form, but I didn't know what was preferred. I will change this.

Reply:
The whole concept of decl strings makes no sense to me and it seems like a lot of OBSProject staff don't really have any idea either, but the way I suggested it is how most procs are implemented

Reply:
Making this change results in the following error in the OBS log:

```
07:01:10.694: [ptz-controls] PTZ controller OBS Plugin"v0.10.3"
07:01:10.694: Errors/warnings for 'void ptz_get_proc_handler(out ptr return)':
07:01:10.694: 
07:01:10.694: declaration (1, 35): 'return' is a reserved name
07:01:10.694: 
07:01:10.694: Function declaration invalid: void ptz_get_proc_handler(out ptr return)
```

I'm reverting to the way I had it before as it seems 'return' cannot be listed as an argument.

Reply:
Up to you. I would have just change to `ret`

Reply:
I would have, except a released version of the face tracker is already using this API